### PR TITLE
Plots for speedup and fingerprints per second

### PR DESCRIPTION
**Pull request by my-alaska**

### CODE REVIEW

File: benchmarking/benchmark.py [hunk 1]
```diff
@@ -54,49 +54,101 @@ def get_times_skfp(X: np.ndarray, transformer_cls: type, **kwargs) -> np.ndarray
     return np.array(result).reshape((len(N_CORES), N_SPLITS))
 
 
-def save_results(
+LEGAL_PLOT_TYPES = ["time", "speedup", "fprints_per_second"]
```
Comment (mjste) [inline benchmarking/benchmark.py:57]: consider ALLOWED_PLOT_TYPES or just PLOT_TYPES

**Review by mjste — CHANGES_REQUESTED** [review]
Add matplotlib to dependencies. For example in eval group

File: benchmarking/benchmark.py (hunk 1 above)
Comment (j-adamczyk) [reply benchmarking/benchmark.py:57]: +1

Comment (j-adamczyk) [reply benchmarking/benchmark.py:57]: Also, maybe `fps_per_second`? We use `fps` variable quite a bit

File: benchmarking/benchmark.py [hunk 2]
```diff
@@ -54,49 +54,101 @@ def get_times_skfp(X: np.ndarray, transformer_cls: type, **kwargs) -> np.ndarray
...
+
+
+def save_plots(
     n_molecules: int,
     times: np.ndarray,
     title: str = "",
     save: bool = True,
+    plot_type: str = "time",
```
Comment (j-adamczyk) [inline benchmarking/benchmark.py:65]: I wouldn't make default value for `plot_type`, it's quite essential after all

File: benchmarking/benchmark.py [hunk 3]
```diff
@@ -54,49 +54,101 @@ def get_times_skfp(X: np.ndarray, transformer_cls: type, **kwargs) -> np.ndarray
...
 
-def save_results(
+LEGAL_PLOT_TYPES = ["time", "speedup", "fprints_per_second"]
+
+
+def save_plots(
     n_molecules: int,
     times: np.ndarray,
     title: str = "",
     save: bool = True,
```
Comment (j-adamczyk) [inline benchmarking/benchmark.py:64]: Why `save` parameter in `save_plots` function? Either it always saves, or change function name, e.g. `make_plot`

Comment (j-adamczyk) [reply benchmarking/benchmark.py:64]: Also it's always a singular plot, isn't it?

File: benchmarking/benchmark.py [hunk 4]
```diff
@@ -54,49 +54,101 @@ def get_times_skfp(X: np.ndarray, transformer_cls: type, **kwargs) -> np.ndarray
...
+    if plot_type == "time":
+        ax1.set_ylabel("Time of computation")
+        for i, y in zip(N_CORES, times):
+            ax1.plot(X, y, marker="o", label=f"# cores: {i}")
+    elif plot_type == "speedup":
+        ax1.set_ylabel("Speedup")
+        for i, y in zip(N_CORES[1:], times[1:]):
+            ax1.plot(X, times[0] / y, marker="o", label=f"# cores: {i}")
+    elif plot_type == "fprints_per_second":
+        ax1.set_ylabel("Fingerprints / s")
```
Comment (j-adamczyk) [inline benchmarking/benchmark.py:86]: Maybe `Fingerprints per second`? I prefer verbose axis names

File: benchmarking/benchmark.py [hunk 5]
```diff
@@ -54,49 +54,101 @@ def get_times_skfp(X: np.ndarray, transformer_cls: type, **kwargs) -> np.ndarray
...
-
-    for i, y in zip(N_CORES, times):
-        ax1.plot(X, y, label=f"our time - # cores: {i}")
-
-    ax1.set_ylabel("Time of computation [s]")
-    ax1.set_xlabel("Number of fingerprints")
+    ax1.set_xlabel("Number of molecules")
+
+    if plot_type == "time":
+        ax1.set_ylabel("Time of computation")
```
Comment (j-adamczyk) [inline benchmarking/benchmark.py:78]: Add unit, e.g. `Time of computation [s]`

File: benchmarking/benchmark.py [hunk 6]
```diff
@@ -54,49 +54,101 @@ def get_times_skfp(X: np.ndarray, transformer_cls: type, **kwargs) -> np.ndarray
...
 
 
-def save_results(
+LEGAL_PLOT_TYPES = ["time", "speedup", "fprints_per_second"]
+
+
+def save_plots(
     n_molecules: int,
     times: np.ndarray,
     title: str = "",
```
Comment (j-adamczyk) [inline benchmarking/benchmark.py:63]: I wouldn't make a default value here. This is pretty essential information, after all

File: benchmarking/benchmark.py [hunk 7]
```diff
@@ -54,49 +54,101 @@ def get_times_skfp(X: np.ndarray, transformer_cls: type, **kwargs) -> np.ndarray
...
 
     plt.close(fig)
 
 
+def save_combined_plot(
+    n_molecules: int,
+    fingerprints: list,
+    times: list,
+    save: bool = True,
+    type: str = "time",
```
Comment (j-adamczyk) [inline benchmarking/benchmark.py:111]: Same comments as above

File: benchmarking/benchmark.py [hunk 8]
```diff
@@ -54,49 +54,101 @@ def get_times_skfp(X: np.ndarray, transformer_cls: type, **kwargs) -> np.ndarray
...
+
+    if type == "time":
+        file_name = "times_of_sequential_computation"
+        ax1.set_xlabel("Time of computation")
+        ax1.set_title("Times of sequential computation for all fingerprints")
+        ax1.barh(fp_names, [time[0, -1] for time in times], color="skyblue")
+    elif type == "speedup":
+        file_name = "speedup_for_all_cores"
+        ax1.set_xlabel("speedup")
+        ax1.set_title("Speedup for all fingerprints")
```
Comment (j-adamczyk) [inline benchmarking/benchmark.py:126]: Note the number of cores: `f"Speedup for {MAX_CORES} cores`

File: benchmarking/benchmark.py [hunk 9]
```diff
@@ -54,49 +54,101 @@ def get_times_skfp(X: np.ndarray, transformer_cls: type, **kwargs) -> np.ndarray
...
+        ax1.barh(
+            fp_names, [time[-1, 0] / time[-1, -1] for time in times], color="skyblue"
+        )
+    elif type == "fprints_per_second":
+        file_name = "fingerprints_per_second_sequential"
+        ax1.set_xlabel("Fingerprints / s")
+        ax1.set_title("Molecules per second for all fingerprints")
```
Comment (j-adamczyk) [inline benchmarking/benchmark.py:133]: Maybe `Molecules processed per second`?

File: benchmarking/benchmark.py [hunk 10]
```diff
@@ -54,49 +54,101 @@ def get_times_skfp(X: np.ndarray, transformer_cls: type, **kwargs) -> np.ndarray
...
+) -> None:
+    fig = plt.figure(figsize=(15, 10))
+    ax1 = fig.add_subplot()
+    ax1.set_ylabel("Fingerprints")
+    fp_names = [fp.__name__ for fp in fingerprints]
+
+    if type == "time":
+        file_name = "times_of_sequential_computation"
+        ax1.set_xlabel("Time of computation")
+        ax1.set_title("Times of sequential computation for all fingerprints")
```
Comment (j-adamczyk) [inline benchmarking/benchmark.py:121]: `Sequential computation time`. The `for all fingerprints` part is not necessary for any plot

File: benchmarking/benchmark.py [hunk 11]
```diff
@@ -54,49 +54,101 @@ def get_times_skfp(X: np.ndarray, transformer_cls: type, **kwargs) -> np.ndarray
...
+            fp_names, [n_molecules / time[0, -1] for time in times], color="skyblue"
+        )
+    else:
+        return
+
+    fig.tight_layout()
+
+    if save:
+        os.makedirs(PLOT_DIR, exist_ok=True)
+        plt.savefig(os.path.join(PLOT_DIR, f"{file_name}.png"))
```
Comment (j-adamczyk) [inline benchmarking/benchmark.py:144]: Format should also be configurable, e.g. LaTeX files work best with PDF plots

File: benchmarking/benchmark.py [hunk 12]
```diff
@@ -145,16 +197,30 @@ def save_results(
         WHIMFingerprint,
     ]
 
-    for fingerprint in fingerprint_constructors:
+    times_for_all_fingerprints = []
```
Comment (j-adamczyk) [inline benchmarking/benchmark.py:200]: Just `all_times` will suffice

**Review by j-adamczyk — CHANGES_REQUESTED** [review]

File: benchmarking/benchmark.py (hunk 3 above)
Comment (my-alaska) [reply benchmarking/benchmark.py:64]: changed to make_plot

**Review by mjste — APPROVED** [review]

**Review by j-adamczyk — APPROVED** [review]

**Pull request merged**